### PR TITLE
Update cucim.skimage API to match scikit-image 0.19

### LINKED ISSUE #98: [FEA] Update API for consistency with upcoming scikit-image 0.19
## New functions introduced in 0.19 that are easy to port to cuCIM (via CuPy)
scikit-image/scikit-image#5158: Add normalized mutual information metric 
scikit-image/scikit-image#5308: New illuminants were added to the color conversions
scikit-image/scikit-image#5382: Added ND butterworth filter
scikit-image/scikit-image#5420: Add no-reference perceptual blur metric. 

Other easy functions are `skimage.feature.blob_dog` and `skimage.feature.blob_log`. Those two existed in v0.18, but had either bug fixes and/or a new argument added in v0.19.

## channel_axis support:
   scikit-image 0.19 adds a `channel_axis` argument that should now be used instead of the `multichannel` boolean. In scikit-image 1.0, the `multichannel` argument will likely be removed. We should start supporting `channel_axis` in cuCIM. Corresponding upstream PRs are:

scikit-image/scikit-image#5228: Decorators for helping with the multichannel->channel_axis transition
scikit-image/scikit-image#5284: multichannel to channel_axis (1 of 6): features and draw
scikit-image/scikit-image#5285: multichannel to channel_axis (2 of 6): transform functions
scikit-image/scikit-image#5286: multichannel to channel_axis (3 of 6): filters
scikit-image/scikit-image#5287: multichannel to channel_axis (4 of 6): metrics and measure
scikit-image/scikit-image#5288: multichannel to channel_axis (5 of 6): restoration
scikit-image/scikit-image#5289: multichannel to channel_axis (6 of 6): segmentation
scikit-image/scikit-image#5462: Add a channel_axis argument to functions in the skimage.color module 
scikit-image/scikit-image#5427: residual multichannel->channel_axis fixes
scikit-image/scikit-image#5348: channel_as_last_axis decorator fix

## Single-precision support

We did much of this already, but need to review for consistency with upstream now that it has been implemented there. Also, can expand the test cases using parameterization over dtypes as was done upstream.

scikit-image/scikit-image#4880: Richardson-Lucy deconvolution: allow single-precision computation
scikit-image/scikit-image#5200: Add float32 support to moments_hu
scikit-image/scikit-image#5204: single precision support in skimage.registration
scikit-image/scikit-image#5219: single precision support in skimage.restoration
scikit-image/scikit-image#5220: single precision support in skimage.metrics
scikit-image/scikit-image#5344: single precision support in moments functions
scikit-image/scikit-image#5353: single precision support in skimage.features
scikit-image/scikit-image#5354: single precision support in skimage.filters
scikit-image/scikit-image#5372: improved single precision support in skimage.transform
scikit-image/scikit-image#5373: single precision support in skimage.segmentation
scikit-image/scikit-image#5443: support single precision in skimage.color

## API/Deprecations
- The ``selem`` argument has been renamed to ``footprint`` throughout the library. The ``footprint`` argument is now deprecated. (scikit-image/scikit-image#5445)
- Deprecate ``in_place`` in favor of the use of an explicit ``out`` argument in ``skimage.morphology.remove_small_objects``,  ``skimage.morphology.remove_small_holes`` and ``skimage.segmentation.clear_border``
- The ``input`` argument of ``skimage.measure.label`` has been renamed ``label_image``. The old name is deprecated.
- standardize on ``num_iter`` for paramters describing the number of iterations and ``max_num_iter`` for parameters specifying an iteration limit.
- The names of several parameters in ``skimage.measure.regionprops`` have been updated so that properties are better grouped by the first word(s) of the name. The old names will continue to work for backwards compatibility.
- In ``measure.label``, the deprecated ``neighbors`` parameter has been removed (use ``connectivity`` instead).
- The deprecated ``skimage.color.rgb2grey`` and ``skimage.color.grey2rgb`` functions have been removed (use ``skimage.color.rgb2gray`` and   ``skimage.color.gray2rgb`` instead).
- ``skimage.color.rgb2gray`` no longer allows grayscale or RGBA inputs.
- The deprecated ``alpha`` parameter of ``skimage.color.gray2rgb`` has now been removed. Use ``skimage.color.gray2rgba`` for conversion to RGBA.
- Attempting to warp a boolean image with ``order > 0`` now raises a ValueError.
- When warping or rescaling boolean images, setting ``anti-aliasing=True`` will raise a ValueError.
- The ``bg_label`` parameter of ``skimage.color.label2rgb`` is now 0.
- The deprecated ``skimage.feature.register_translation`` function has been removed (use ``skimage.registration.phase_cross_correlation`` instead).
- The deprecated ``skimage.feature.masked_register_translation`` function has been removed (use ``skimage.registration.phase_cross_correlation`` instead).
- The default ``mode`` in ``skimage.filters.hessian`` is now ``'reflect'``.
- The default boundary ``mode`` in ``skimage.filters.sato`` is now ``'reflect'``.

## Bug fixes
- Input ``labels`` argument renumbering in ``skimage.feature.peak_local_max`` is avoided (scikit-image/scikit-image#5047).
- Nonzero values at the image edge are no longer incorrectly marked as a boundary when using ``find_bounaries`` with mode='subpixel' (scikit-image/scikit-image#5447).
- Fix return dtype of ``_label2rgb_avg`` function.
- Ensure ``skimage.color.separate_stains`` does not return negative values.
- Handle 1D arrays properly in ``skimage.filters.gaussian``.
- Fix Laplacian matrix size bug in ``skimage.segmentation.random_walker``.
- Regionprops table (``skimage.measure.regionprops_table``) dtype bugfix.
- Fix ``skimage.transform.rescale`` when using a small scale factor.
- Fix multichannel ``intensity_image`` extra_properties in regionprops.
- Fix error message for ``skimage.metric.structural_similarity`` when image is too small.
- Do not mark image edges in 'subpixel' mode of ``skimage.segmentation.find_boundaries``.
- Fix behavior of ``skimage.exposure.is_low_contrast`` for boolean inputs.
- Fix wrong syntax for the string argument of ValueError in ``skimage.metric.structural_similarity`` .
- Fixed NaN issue in ``skimage.filters.threshold_otsu``.
- Fix ``skimage.feature.blob_dog`` docstring example and normalization.
- Fix uint8 overflow in ``skimage.exposure.adjust_gamma``.
- Fix broken doctests in ``skimage.exposure.histogram`` and  ``skimage.measure.regionprops_table``. (scikit-image/scikit-image#5522)
- Corrected phase correlation in ``skimage.register.phase_cross_correlation``.  (scikit-image/scikit-image#5461)

### PR DESCRIPTION
**Pull request by grlee77**
This PR builds on top of #189 and introduces new deprecations as in scikit-image 0.19.

closes #98

Specifically:
 `selem` -> `footprint`
`grey` -> `gray`
`iterations` -> `num_iter`
`max_iter` -> `max_num_iter`
`min_iter` -> `min_num_iter`

### CODE REVIEW

**Review by Ethyling — APPROVED** [review]

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]
We'll also need to get the `scikit-image` version updated in the `integration` repository below. @grlee77, can you open up a PR for this?

https://github.com/rapidsai/integration/blob/f0848c2c340de9e511c72d6edcc66e2b310f1c9e/conda/recipes/versions.yaml#L147-L148

**Comment by grlee77** [comment]
@jakirkham , @gigony : I think I now have all the updates from scikit-image 0.19 in this one PR. Aside from the basic updates, I did add four new functions introduced in 0.19 that were possible purely with the CuPy API:

`skimage.filters.butterworth`
`skimage.measure.blur_effect`
`skimage.metrics.normalized_mutual_information`
`skimage.transform.resize_local_mean`

Aside from the updates to 0.19, I also ran `pytest --doctest-modules` and fixed any failing doctests. It should be possible to test on CI with scikit-image 0.18 or 0.19 (We mainly use scikit-image to retrieve test data that is the same in both versions).

**Comment by jakirkham** [comment]
> We'll also need to get the `scikit-image` version updated in the `integration` repository below. @grlee77, can you open up a PR for this?
> 
> https://github.com/rapidsai/integration/blob/f0848c2c340de9e511c72d6edcc66e2b310f1c9e/conda/recipes/versions.yaml#L147-L148

@ajschmidt8, done in PR ( https://github.com/rapidsai/integration/pull/414 ) :)

**Comment by grlee77** [comment]
tests are close to passing now. It looks like there was one test case in `src/cucim/skimage/restoration/tests/test_j_invariant.py` that assumed scikit-image 0.19 was installed. I can update that to work for either version.

There were a handful of other failures that look unrelated to the PR. e.g. for the centos7 case I saw
```
00:42:03 tests/unit/clara/test_image_cache.py:18: in test_get_nocache
00:42:03     from cucim import CuImage
00:42:03 E   ImportError: cannot import name 'CuImage' from 'cucim' (/workspace/python/cucim/src/cucim/__init__.py)
```

**Comment by jakirkham** [comment]
> There were a handful of other failures that look unrelated to the PR. e.g. for the centos7 case I saw
> 
> ```
> 00:42:03 tests/unit/clara/test_image_cache.py:18: in test_get_nocache
> 00:42:03     from cucim import CuImage
> 00:42:03 E   ImportError: cannot import name 'CuImage' from 'cucim' (/workspace/python/cucim/src/cucim/__init__.py)
> ```

So I opened PR ( https://github.com/rapidsai/cucim/pull/198 ) to test CI with a trivial blank line addition. However it did not show the same failures. Perhaps there is more still going on here?

**Comment by grlee77** [comment]
> So I opened PR ( #198 ) to test CI with a trivial blank line addition. However it did not show the same failures. Perhaps there is more still going on here?

I think the PR was made against v22.02.00, but I must have started this on v21.12.00, because `git describe --abbrev=0 --tags` was giving `v21.12.00` and not `v22.02.00a`. I rebased and force-pushed the branch and hopefully it will work this time around...

**Comment by jakirkham** [comment]
Looks like we are still seeing the issue on CI.

Noticed the PR includes some tricks around lazy loading, defining `__dir__` in the top-level module, etc. Is it possible something like this is causing issues?

Saw some changes around `pytest` setup and teardown as well. Wouldn't think that would be related, but that might be another avenue to explore

**Comment by jakirkham** [comment]
@gigony would you be able to take a look as well? There are some `import` changes that it would be good to confirm work for you 🙂

**Comment by grlee77** [comment]
Okay, it passed now. I susupect it was possibly the `conftest.py`, but did not test with that change in isolation. We do not necessarily need that file here (We use it in one CI case on scikit-image where it is set to fail if any unexpected warnings are encountered).

@gigony, would specifically want to know if you think we should keep the change in https://github.com/rapidsai/cucim/pull/190/commits/c8177c02969f410aa79d7051face474101af77c1 that added lazy loading to the top-level cucim. If not we can revert the change to that __init__.py and just have it for the skimage subpackage. Basically the primary benefit is that users only need to import top-level `cucim` and could then directly call things like `cucim.skimage.filters.sobel` without having to do a separate `import cucim.skimage.filters` first. That change was made for scikit-image in 0.19, but we don't necesarilly have to extend it up from `cucim.skimage` to the top-level `cucim` module.

**Comment by grlee77** [comment]
One other option we could potentially do on CI is to run one case with environment variable  `EAGER_IMPORT=1` defined. That can be used to test with eager importing of everything (i.e. disables lazy loading). That can help rule out potential issues with circular imports.

File: python/cucim/src/cucim/skimage/metrics/tests/test_structural_similarity.py [hunk 1]
```diff
@@ -115,30 +124,51 @@ def test_structural_similarity_multichannel():
...
+    # replicate across three channels.  should get identical value
+    Xc = cp.tile(X[..., cp.newaxis], (1, 1, 3))
+    Yc = cp.tile(Y[..., cp.newaxis], (1, 1, 3))
+    with expected_warnings(["`multichannel` is a deprecated argument"]):
+        S2 = structural_similarity(Xc, Yc, multichannel=True, win_size=3)
+    assert_almost_equal(S1, S2)
 
 
-@pytest.mark.parametrize('dtype', [cp.uint8, cp.float32, cp.float64])
+@pytest.mark.parametrize('dtype', [np.uint8, np.float32, np.float64])
```
Comment (gigony) [inline python/cucim/src/cucim/skimage/metrics/tests/test_structural_similarity.py:171]: I wonder why some code is changed to `cp.uint8` and some other code is changed to `np.uint8`.

**Review by gigony — APPROVED** [review]
Looks good to me!

It was good to see how lazy module works :)  

Left minor comments.

**Comment by gigony** [comment]
> @gigony, would specifically want to know if you think we should keep the change in c8177c0 that added lazy loading to the top-level cucim (I accidentally commited that in the same commit with the flake8 fix). If not we can revert the change to cucim/__init__.py and just have lazy loading within the cucim.skimage subpackage. Basically the primary benefit is that users only need to import top-level cucim and could then directly call things like cucim.skimage.filters.sobel without having to do a separate import cucim.skimage.filters first. That change was made for scikit-image in 0.19, but we don't necesarilly have to extend it up from cucim.skimage to the top-level cucim module.

I think it is okay and having it would make it easier to restruct modules under `clara` or apply lazy imports in the future.
Thanks for the update!

File: python/cucim/src/cucim/skimage/filters/_gabor.py [hunk 2]
```diff
@@ -81,34 +83,44 @@ def gabor_kernel(frequency, theta=0, bandwidth=1, sigma_x=None, sigma_y=None,
     if sigma_y is None:
         sigma_y = _sigma_prefactor(bandwidth) / frequency
 
+    if float_dtype is not None:
+        if dtype is not None:
+            raise ValueError(
+                "Cannot set both dtype and float_dtype. Please dtype instead "
```
[the remark this answers is not in the document]
Comment (grlee77) [reply python/cucim/src/cucim/skimage/filters/_gabor.py]: Thanks @gigony, I updated this as suggested

File: python/cucim/src/cucim/skimage/metrics/tests/test_structural_similarity.py (hunk 1 above)
Comment (grlee77) [reply python/cucim/src/cucim/skimage/metrics/tests/test_structural_similarity.py]: updated in 8438d64

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
@gpucibot merge

**Pull request merged**